geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/ReceiptTimeScheme.scala

```diff
@@ -0,0 +1,294 @@
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.fs.storage.common.partitions
+
+import org.geotools.filter.visitor.DuplicatingFilterVisitor
+import org.geotools.temporal.`object`.{DefaultInstant, DefaultPeriod, DefaultPosition}
+import org.locationtech.geomesa.fs.storage.api.PartitionScheme.SimplifiedFilter
+import org.locationtech.geomesa.fs.storage.api.{NamedOptions, PartitionScheme, PartitionSchemeFactory}
+import org.locationtech.geomesa.fs.storage.common.partitions.ReceiptTimeScheme.BufferingFilterVisitor
+import org.locationtech.geomesa.utils.geotools.converters.FastConverter
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+import org.opengis.filter.expression.{Expression, Literal, PropertyName}
+import org.opengis.filter.temporal.{After, Before, During, TEquals}
+import org.opengis.temporal.{Instant, Period}
+
+import java.util.Date
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.Duration
+
+/**
+ * Scheme for partitioning based on "receipt time", i.e. when a message is received. Generally this is useful
+ * only for reading existing data that may have been aggregated and stored by an external process.
+ *
+ * @param delegate delegate date time scheme options
+ * @param buffer amount of time to buffer queries by, in order to match a feature date to a receipt time date -
+ *               i.e. the amount of latency in the ingest process
+ */
+case class ReceiptTimeScheme(delegate: DateTimeScheme, buffer: Duration) extends PartitionScheme {
+
+  override val depth: Int = delegate.depth
+
+  override val pattern: String = delegate.pattern
+
+  override def getPartitionName(feature: SimpleFeature): String = delegate.getPartitionName(feature)
+
+  override def getSimplifiedFilters(filter: Filter, partition: Option[String]): Option[Seq[SimplifiedFilter]] = {
+    delegate.getSimplifiedFilters(buffered(filter), partition).map { filters =>
+      // always use the full filter since our dates are not guaranteed to match the partition bounds
+      filters.map(f => f.copy(filter = filter))
+    }
+  }
+
+  override def getIntersectingPartitions(filter: Filter): Option[Seq[String]] =
+    delegate.getIntersectingPartitions(buffered(filter))
+
+  override def getCoveringFilter(partition: String): Filter =
+    throw new NotImplementedError("Dates may overlap in multiple partitions")
+
+  private def buffered(filter: Filter): Filter =
+    filter.accept(new BufferingFilterVisitor(buffer, delegate.dtg), null).asInstanceOf[Filter]
+}
+
+object ReceiptTimeScheme {
+
+  val Name = "receipt-time"
+
+  object Config {
+    val DateTimeSchemaOpt: String = "datetime-scheme"
+    val BufferOpt        : String = "buffer"
+  }
+
+  class ReceiptTimePartitionSchemeFactory extends PartitionSchemeFactory {
```

Comment:
Perhaps we should extend the `DateTimePartitionSchemeFactory` class instead, since the receipt-time partitioning scheme is a special case of the date-time scheme?

Reply:
yeah, we could - i think it would mostly just change the call from `PartitionSchemeFactory.load` to `super.load`, but it seems like we would also want to make the change you suggested above in making the scheme extend date time scheme, to be consistent. i think that would have been a valid way to go, but i don't see a compelling reason to change it now.